### PR TITLE
[JAX] Flax params initialization with weight_dtype

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description
[UPDATED] PR #1472 caused convergence issues in MaxText. 

MaxText has `weight_dtype` as the data type of the parameters and `dtype` as the compute type, whereas the current TE only has `dtype`.

This PR adds the `weight_dtype` as an optional arg with default value `jnp.float32`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by huanghua1994** [comment]
Just a note here: MaxText initialize some parameters with `weight_dtype` then cast them into `dtype`, these two dtypes are not always the same. And `weight_dtype` may or may not be `jnp.float32`.

**Comment by phu0ngng** [comment]
Good point! I added `init_dtype` as an optional arg for all the modules.
All the initializers will take `init_dtype` with `jnp.float32` by default.

File: transformer_engine/jax/flax/transformer.py [hunk 1]
```diff
@@ -149,8 +150,8 @@ def __call__(
         del self.scale_factor
 
         if self.float32_logits:
-            query = query.astype(jnp.float32)
-            key = key.astype(jnp.float32)
+            query = query.astype(self.dtype)
+            key = key.astype(self.dtype)
```
Comment (phu0ngng) [inline transformer_engine/jax/flax/transformer.py:154]: @zlsh80826 Please confirm if this is the right fix.

Comment (zlsh80826) [reply transformer_engine/jax/flax/transformer.py:154]: no, if specifying self.float32_logits, then the user is expected the computation become float32, that's for the numerical stability of the following softmax.

Comment (zlsh80826) [inline transformer_engine/jax/flax/transformer.py:154]:
```suggestion
            query = query.astype(jnp.float32)
            key = key.astype(jnp.float32)
```

**Review by huanghua1994 — APPROVED** [review]
